sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,984 @@
+            The number of OpenMP threads to use for the reduction.
+            Parallelism is done on chunks and the sharding of chunks
+            depends on the `strategy` set on
+            :method:`~PairwiseDistancesArgKmin.compute`.
+
+            None and -1 means using all processors.
+
+        Returns
+        -------
+        argkmin: PairwiseDistancesArgKmin
+            The suited PairwiseDistancesArgKmin implementation.
+        """
+        # This factory comes to handle specialisations.
+        if metric in ("fast_euclidean", "fast_sqeuclidean") and not issparse(X) and not issparse(Y):
+            use_squared_distances = metric == "fast_sqeuclidean"
+            return FastEuclideanPairwiseDistancesArgKmin(
+                X=X, Y=Y, k=k,
+                use_squared_distances=use_squared_distances,
+                chunk_size=chunk_size,
+                metric_kwargs=metric_kwargs,
+            )
+
+        return PairwiseDistancesArgKmin(
+            datasets_pair=DatasetsPair.get_for(X, Y, metric, metric_kwargs),
+            k=k,
+            chunk_size=chunk_size,
+        )
+
+    def __init__(
+        self,
+        DatasetsPair datasets_pair,
+        ITYPE_t k,
+        chunk_size=None,
+        n_threads=None,
+    ):
+        super().__init__(datasets_pair, chunk_size, n_threads)
+
+        self.k = check_scalar(k, "k", Integral, min_val=1)
+
+        # Allocating pointers to datastructures but not the datastructures themselves.
+        # There are as many pointers as available threads.
+        # However, when reducing on small datasets, there can be more pointers than
+        # actual threads.
+        # In this case, some pointers will be dynamically allocated but there won't
+        # be allocated yet unused data-structures referenced by them.
+        #
+        # For the sake of explicitness:
+        #   - when parallelizing on X, those heaps pointers are referencing
+        #   (with proper offsets) addresses of the two main heaps (see bellow)
+        #   - when parallelizing on Y, those heaps pointer heaps are referencing
+        #   small heaps which are thread-wise-allocated and whose content will be
+        #   merged with the main heaps'.
+        self.heaps_r_distances_chunks = <DTYPE_t **> malloc(
+            sizeof(DTYPE_t *) * self.effective_omp_n_thread
+        )
+        self.heaps_indices_chunks = <ITYPE_t **> malloc(
+            sizeof(ITYPE_t *) * self.effective_omp_n_thread
+        )
+
+        # Main heaps used by PairwiseDistancesArgKmin.compute to return results.
+        self.argkmin_indices = np.full((self.n_samples_X, self.k), 0, dtype=ITYPE)
+        self.argkmin_distances = np.full((self.n_samples_X, self.k), DBL_MAX, dtype=DTYPE)
+
+    def __dealloc__(self):
+        if self.heaps_indices_chunks is not NULL:
+            free(self.heaps_indices_chunks)
+
+        if self.heaps_r_distances_chunks is not NULL:
+            free(self.heaps_r_distances_chunks)
```

Comment:
I think it would be nice to bind the life cycle of everything allocated by this object with this object, kind of like [RAII](https://en.cppreference.com/w/cpp/language/raii).

Can we determine the parallel strategy during `__cinit__` and malloc everything in `__cinit__`? This way we can also free everything during `__dealloc__`. I am not sure if that plays well with openmp and cpu data locality. 

If we have to keep the mallocs in `compute`, I think it would still be better to free everything in `__dealloc__`. It would need to condition on strategy because nothing is malloced during `parallel_on_X`.

Reply:
Yes, I also do not think that `compute` is properly design because it might have some side effect if called more than once (this is not the case currently in the code base, but it might be).

I think such changes might also prevent subsequent mistakes when extending this class for new reductions (which is a probable subsequent work). I will inspect the difference it in details. I hope it can play well with data locality.

@jeremiedbb also made a similar remark regarding merge some logic made in `compute` within `__(c)init__`.

Reply:
@jeremiedbb and @thomasjpfan: I think that I need your thoughts here about the best way to proceed.

Reply:
In terms of UX, this interface gets used as follows:

```python
pdr = PairwiseDistancesArgKmin.get_for(
    X=X, Y=Y, k=k, metric=metric, metric_kwargs=metric_kwargs
)
values, indices = pdr.compute(strategy="auto", return_distance=True)
```

As we do not need the `pdr` instance, probably we could solve this RAII problem using a sole interface, e.g.:
```python
values, indices = PairwiseDistancesArgKmin.get_for(
    X=X, Y=Y, k=k,
    metric=metric,
    metric_kwargs=metric_kwargs,
    strategy="auto",
    return_distance=True
)
```

What do you think?

Reply:
That's an option. This this case, this new / extended `get_for` should probably be renamed to `compute` and we might want to keep a private `_get_datasetpairs_for` internally?

Reply:
With `PairwiseDistancesArgKmin.compute` it becomes a class function with no state. Internally, we would still need to create a `PairwiseDistancesArgKminResults` class to store the state, `argkmin_indices`, `heaps` etc.

I feel like this is a bit cleaner compared having a stateful `PairwiseDistancesArgKmin`. This way, `PairwiseDistancesArgKmin` is the "pure functional form" that computes, and `PairwiseDistancesArgKminResults` is a simple "dataclass" which stores the results.

Reply:
> This this case, this new / extended get_for should probably be renamed to compute

Exploring it now.

> we might want to keep a private _get_datasetpairs_for internally

To me, this is uneeded and YAGNI for now. As this code is private, this can always be added later if we need to.

> With PairwiseDistancesArgKmin.compute it becomes a class function with no state. Internally, we would still need to create a PairwiseDistancesArgKminResults class to store the state, argkmin_indices, heaps etc.
> 
> I feel like this is a bit cleaner compared having a stateful PairwiseDistancesArgKmin. This way, PairwiseDistancesArgKmin is the "pure functional form" that computes, and PairwiseDistancesArgKminResults is a simple "dataclass" which stores the results.

I agree.

I'll start with the design I proposed (by keeping an instance of `PairwiseDistancesArgKmin` for `PairwiseDistancesArgKmin.compute` internally). From there, `PairwiseDistancesArgKminResults` (the wrapper proposed by @thomasjpfan) can be introduced more easily. To me, I believe we can proceed without this wrapper, let's see.

This is to be explored in https://github.com/jjerphan/scikit-learn/tree/pairwise-distances-argkmin-raii

Reply:
jjerphan#5 looks great. I do not see the need to introduce `PairwiseDistancesArgKminResults` in particular because the heaps are temporary datastructures only used internally during the computation, not "results" per-se.